Re-enable tests on Jenkins Release pipelines

### Component

JKube Kit

### Task description

I had to temporarily disable test in the release pipelines because the tests related to the remote resource fragments features are failing:

```
Caused by: java.io.IOException: Failed to download URL https://gist.githubusercontent.com/lordofthejars/ac2823cec7831697d09444bbaa76cd50/raw/e4b43f1b6494766dfc635b5959af7730c1a58a93/deployment.yaml to  /tmp/jkube6633473296172234022/deployment.yaml: java.util.concurrent.ExecutionException: java.net.UnknownHostException: gist.githubusercontent.com
...
[ERROR] Errors: 
[ERROR]   KubernetesHelperTest.testGetResourceFragmentFromSourceWithNullResourceDirAndSomeRemotes:115 » IllegalArgument
[ERROR]   KubernetesHelperTest.testGetResourceFragmentFromSourceWithSomeResourceDirAndSomeRemotes:130 » IllegalArgument
[ERROR]   KubernetesHelperTest.testResourceFragmentsWithRemotes:82 » IllegalArgument jav...
```

We need to refactor these tests so that the fragments are downloaded from a temporary localhost web server.

Once these tests are replaced, the pipelines should be modified again to re-enable the tests.

- [x] Refactor tests
- [x] Re-enable tests in pipelines

It's not just KubernetesHelperTest that's failing on jenkins. Ever since https://github.com/eclipse/jkube/pull/1878 got merged our [daily jenkins build](https://ci.eclipse.org/jkube/job/daily-build/) has been failing :

```
[ERROR] Failures: 
[ERROR]   DebugServiceTest.debugWithApplicableEntities:212 
[ERROR]   DebugServiceTest.debugWithApplicableEntitiesAndSuspend
```

--- a/jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/TestHttpStaticServer.java
+++ b/jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/TestHttpStaticServer.java
@@ -13,8 +13,8 @@
  */
 package org.eclipse.jkube.kit.common;
 
-import com.sun.net.httpserver.Headers;
 import com.sun.net.httpserver.HttpServer;
+import org.apache.commons.io.IOUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -24,54 +24,75 @@ import java.io.FileInputStream;
 import java.io.IOException;
 import java.io.OutputStream;
 import java.net.InetSocketAddress;
-import java.util.concurrent.ExecutorService;
-import java.util.concurrent.Executors;
+import java.net.URLConnection;
+import java.util.Optional;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static org.eclipse.jkube.kit.common.util.AsyncUtil.async;
+import static org.eclipse.jkube.kit.common.util.IoUtil.getFreeRandomPort;
 
 public class TestHttpStaticServer implements Closeable {
-  private final HttpServer server;
-  private final File staticDirectory;
+
   private static final Logger log = LoggerFactory.getLogger(TestHttpStaticServer.class);
+  private final CompletableFuture<HttpServer> server;
 
-  public TestHttpStaticServer(int port, File staticDirectory) throws IOException {
-    server = HttpServer.create(new InetSocketAddress(port), 0);
-    this.staticDirectory = staticDirectory;
-    ExecutorService executorService = Executors.newSingleThreadExecutor();
-    executorService.submit(this::startServer);
-  }
-
-  private void startServer() {
-    server.createContext("/", exchange -> {
-      File file = new File(staticDirectory, exchange.getRequestURI().getPath()).getCanonicalFile();
-
-      if (file.isFile()) {
-        Headers headers = exchange.getResponseHeaders();
-        headers.set("Content-Type", "text/html");
-        exchange.sendResponseHeaders(200, 0);
-
-        try (OutputStream outputStream = exchange.getResponseBody();
-             FileInputStream fis = new FileInputStream(file)) {
-          final byte[] buffer = new byte[0x10000];
-          int count;
-          while ((count = fis.read(buffer)) >= 0) {
-            outputStream.write(buffer, 0, count);
-          }
-        }
-      } else {
-        String response = "404 (Not Found)\n";
-        exchange.sendResponseHeaders(404 , response.length());
-        try (OutputStream os = exchange.getResponseBody()) {
-          os.write(response.getBytes());
-        }
-      }
-    });
-    server.setExecutor(null);
-    log.info("Starting server");
-    server.start();
+  public TestHttpStaticServer(File staticDirectory) {
+    server = async(startServer(staticDirectory));
   }
 
   @Override
   public void close() throws IOException {
     log.info("Stopping server");
-    server.stop(0);
+    getServer().stop(0);
+  }
+
+  private static Callable<HttpServer> startServer(File staticDirectory) {
+    return () -> {
+      final int port = getFreeRandomPort();
+      final HttpServer ret = HttpServer.create(new InetSocketAddress(port), 0);
+      ret.createContext("/", exchange -> {
+        final File file = new File(staticDirectory, exchange.getRequestURI().getPath()).getCanonicalFile();
+        if (file.isFile()) {
+
+          try (
+            OutputStream outputStream = exchange.getResponseBody();
+            FileInputStream fis = new FileInputStream(file)
+          ) {
+            exchange.getResponseHeaders().set("Content-Type",
+              Optional.ofNullable(URLConnection.guessContentTypeFromStream(fis)).orElse("application/octet-stream"));
+            exchange.sendResponseHeaders(200, 0);
+            IOUtils.copy(fis, outputStream);
+          }
+        } else {
+          String response = "404 (Not Found)\n";
+          exchange.sendResponseHeaders(404 , response.length());
+          try (OutputStream os = exchange.getResponseBody()) {
+            os.write(response.getBytes());
+          }
+        }
+      });
+      log.info("Starting server");
+      ret.start();
+      return ret;
+    };
+  }
+
+  public int getPort() {
+    return getServer().getAddress().getPort();
+  }
+
+  private HttpServer getServer() {
+    try {
+      return server.get(10, TimeUnit.SECONDS);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
+    } catch (TimeoutException | ExecutionException e) {
+      throw new RuntimeException(e);
+    }
   }
 }

--- a/jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/TestHttpStaticServer.java
+++ b/jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/TestHttpStaticServer.java
@@ -1,0 +1,77 @@
+/**
+ * Copyright (c) 2019 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at:
+ *
+ *     https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.jkube.kit.common;
+
+import com.sun.net.httpserver.Headers;
+import com.sun.net.httpserver.HttpServer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.InetSocketAddress;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+public class TestHttpStaticServer implements Closeable {
+  private final HttpServer server;
+  private final File staticDirectory;
+  private static final Logger log = LoggerFactory.getLogger(TestHttpStaticServer.class);
+
+  public TestHttpStaticServer(int port, File staticDirectory) throws IOException {
+    server = HttpServer.create(new InetSocketAddress(port), 0);
+    this.staticDirectory = staticDirectory;
+    ExecutorService executorService = Executors.newSingleThreadExecutor();
+    executorService.submit(this::startServer);
+  }
+
+  private void startServer() {
+    server.createContext("/", exchange -> {
+      File file = new File(staticDirectory, exchange.getRequestURI().getPath()).getCanonicalFile();
+
+      if (file.isFile()) {
+        Headers headers = exchange.getResponseHeaders();
+        headers.set("Content-Type", "text/html");
+        exchange.sendResponseHeaders(200, 0);
+
+        try (OutputStream outputStream = exchange.getResponseBody();
+             FileInputStream fis = new FileInputStream(file)) {
+          final byte[] buffer = new byte[0x10000];
+          int count;
+          while ((count = fis.read(buffer)) >= 0) {
+            outputStream.write(buffer, 0, count);
+          }
+        }
+      } else {
+        String response = "404 (Not Found)\n";
+        exchange.sendResponseHeaders(404 , response.length());
+        try (OutputStream os = exchange.getResponseBody()) {
+          os.write(response.getBytes());
+        }
+      }
+    });
+    server.setExecutor(null);
+    log.info("Starting server");
+    server.start();
+  }
+
+  @Override
+  public void close() throws IOException {
+    log.info("Stopping server");
+    server.stop(0);
+  }
+}

--- a/jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/KubernetesHelperTest.java
+++ b/jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/KubernetesHelperTest.java
@@ -48,11 +48,13 @@ import io.fabric8.kubernetes.api.model.apps.StatefulSetBuilder;
 import io.fabric8.kubernetes.api.model.batch.v1.JobBuilder;
 import io.fabric8.openshift.api.model.DeploymentConfigBuilder;
 import io.fabric8.openshift.api.model.Template;
+import org.eclipse.jkube.kit.common.TestHttpStaticServer;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.entry;
+import static org.eclipse.jkube.kit.common.util.IoUtil.getFreeRandomPort;
 
 class KubernetesHelperTest {
 
@@ -73,18 +75,22 @@ class KubernetesHelperTest {
     }
 
     @Test
-    void testResourceFragmentsWithRemotes() {
-        // Given
-        List<String> remoteStrList = getRemoteFragments();
-        File localResourceDir = new File(getClass().getResource("/util/fragments").getPath());
+    void testResourceFragmentsWithRemotes() throws IOException {
+        int port = getFreeRandomPort();
+        File remoteDirectory = new File(getClass().getResource("/remote-resources").getFile());
+        try (TestHttpStaticServer ignore = new TestHttpStaticServer(port, remoteDirectory)) {
+            // Given
+            List<String> remoteStrList = getRemoteFragments(port);
+            File localResourceDir = new File(getClass().getResource("/util/fragments").getPath());
 
-        // When
-        File[] fragments = KubernetesHelper.listResourceFragments(remoteStrList, logger, localResourceDir);
+            // When
+            File[] fragments = KubernetesHelper.listResourceFragments(remoteStrList, logger, localResourceDir);
 
-        // Then
-        assertLocalFragments(fragments, 4);
-        assertThat(Arrays.stream(fragments).anyMatch( f -> f.getName().equals("deployment.yaml"))).isTrue();
-        assertThat(Arrays.stream(fragments).anyMatch( f -> f.getName().equals("sa.yml"))).isTrue();
+            // Then
+            assertLocalFragments(fragments, 4);
+            assertThat(Arrays.stream(fragments).anyMatch( f -> f.getName().equals("deployment.yaml"))).isTrue();
+            assertThat(Arrays.stream(fragments).anyMatch( f -> f.getName().equals("sa.yml"))).isTrue();
+        }
     }
 
     @Test
@@ -107,32 +113,40 @@ class KubernetesHelperTest {
     }
 
     @Test
-    void testGetResourceFragmentFromSourceWithNullResourceDirAndSomeRemotes() {
-        // Given
-        List<String> remotes = getRemoteFragments();
+    void testGetResourceFragmentFromSourceWithNullResourceDirAndSomeRemotes() throws IOException {
+        int port = getFreeRandomPort();
+        File remoteDirectory = new File(getClass().getResource("/remote-resources").getFile());
+        try (TestHttpStaticServer ignore = new TestHttpStaticServer(port, remoteDirectory)) {
+            // Given
+            List<String> remotes = getRemoteFragments(port);
 
-        // When
-        File fragmentFile = KubernetesHelper.getResourceFragmentFromSource(null, remotes, "deployment.yaml", logger);
+            // When
+            File fragmentFile = KubernetesHelper.getResourceFragmentFromSource(null, remotes, "deployment.yaml", logger);
 
-        // Then
-        assertThat(fragmentFile).isNotNull()
+            // Then
+            assertThat(fragmentFile).isNotNull()
                 .exists()
                 .hasName("deployment.yaml");
+        }
     }
 
     @Test
-    void testGetResourceFragmentFromSourceWithSomeResourceDirAndSomeRemotes() {
-        // Given
-        File localResourceDir = new File(getClass().getResource("/util/fragments").getPath());
-        List<String> remotes = getRemoteFragments();
+    void testGetResourceFragmentFromSourceWithSomeResourceDirAndSomeRemotes() throws IOException {
+        int port = getFreeRandomPort();
+        File remoteDirectory = new File(getClass().getResource("/remote-resources").getFile());
+        try (TestHttpStaticServer ignore = new TestHttpStaticServer(port, remoteDirectory)) {
+            // Given
+            File localResourceDir = new File(getClass().getResource("/util/fragments").getPath());
+            List<String> remotes = getRemoteFragments(port);
 
-        // When
-        File fragmentFile = KubernetesHelper.getResourceFragmentFromSource(localResourceDir, remotes, "sa.yml", logger);
+            // When
+            File fragmentFile = KubernetesHelper.getResourceFragmentFromSource(localResourceDir, remotes, "sa.yml", logger);
 
-        // Then
-        assertThat(fragmentFile).isNotNull()
+            // Then
+            assertThat(fragmentFile).isNotNull()
                 .exists()
                 .hasName("sa.yml");
+        }
     }
 
     @Test
@@ -414,11 +428,11 @@ class KubernetesHelperTest {
         return envVarList;
     }
 
-    private List<String> getRemoteFragments() {
+    private List<String> getRemoteFragments(int port) {
         List<String> remoteStrList = new ArrayList<>();
 
-        remoteStrList.add("https://gist.githubusercontent.com/lordofthejars/ac2823cec7831697d09444bbaa76cd50/raw/e4b43f1b6494766dfc635b5959af7730c1a58a93/deployment.yaml");
-        remoteStrList.add("https://gist.githubusercontent.com/rohanKanojia/c4ac4ae5533f0bf0dd77d13c905face7/raw/8a7de1e27c1f437c1ccbd186ed247efd967953ee/sa.yml");
+        remoteStrList.add(String.format("http://localhost:%d/deployment.yaml", port));
+        remoteStrList.add(String.format("http://localhost:%d/sa.yml", port));
         return remoteStrList;
     }
 }

--- a/jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/KubernetesHelperTest.java
+++ b/jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/KubernetesHelperTest.java
@@ -54,7 +54,6 @@ import org.junit.jupiter.api.Test;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.entry;
-import static org.eclipse.jkube.kit.common.util.IoUtil.getFreeRandomPort;
 
 class KubernetesHelperTest {
 
@@ -75,12 +74,11 @@ class KubernetesHelperTest {
     }
 
     @Test
-    void testResourceFragmentsWithRemotes() throws IOException {
-        int port = getFreeRandomPort();
+    void resourceFragmentsWithRemotes() throws IOException {
         File remoteDirectory = new File(getClass().getResource("/remote-resources").getFile());
-        try (TestHttpStaticServer ignore = new TestHttpStaticServer(port, remoteDirectory)) {
+        try (TestHttpStaticServer http = new TestHttpStaticServer(remoteDirectory)) {
             // Given
-            List<String> remoteStrList = getRemoteFragments(port);
+            List<String> remoteStrList = getRemoteFragments(http.getPort());
             File localResourceDir = new File(getClass().getResource("/util/fragments").getPath());
 
             // When
@@ -113,12 +111,11 @@ class KubernetesHelperTest {
     }
 
     @Test
-    void testGetResourceFragmentFromSourceWithNullResourceDirAndSomeRemotes() throws IOException {
-        int port = getFreeRandomPort();
+    void getResourceFragmentFromSourceWithNullResourceDirAndSomeRemotes() throws IOException {
         File remoteDirectory = new File(getClass().getResource("/remote-resources").getFile());
-        try (TestHttpStaticServer ignore = new TestHttpStaticServer(port, remoteDirectory)) {
+        try (TestHttpStaticServer http = new TestHttpStaticServer(remoteDirectory)) {
             // Given
-            List<String> remotes = getRemoteFragments(port);
+            List<String> remotes = getRemoteFragments(http.getPort());
 
             // When
             File fragmentFile = KubernetesHelper.getResourceFragmentFromSource(null, remotes, "deployment.yaml", logger);
@@ -131,13 +128,12 @@ class KubernetesHelperTest {
     }
 
     @Test
-    void testGetResourceFragmentFromSourceWithSomeResourceDirAndSomeRemotes() throws IOException {
-        int port = getFreeRandomPort();
+    void getResourceFragmentFromSourceWithSomeResourceDirAndSomeRemotes() throws IOException {
         File remoteDirectory = new File(getClass().getResource("/remote-resources").getFile());
-        try (TestHttpStaticServer ignore = new TestHttpStaticServer(port, remoteDirectory)) {
+        try (TestHttpStaticServer http = new TestHttpStaticServer(remoteDirectory)) {
             // Given
             File localResourceDir = new File(getClass().getResource("/util/fragments").getPath());
-            List<String> remotes = getRemoteFragments(port);
+            List<String> remotes = getRemoteFragments(http.getPort());
 
             // When
             File fragmentFile = KubernetesHelper.getResourceFragmentFromSource(localResourceDir, remotes, "sa.yml", logger);

--- a/jkube-kit/common/src/test/resources/remote-resources/deployment.yaml
+++ b/jkube-kit/common/src/test/resources/remote-resources/deployment.yaml
@@ -1,0 +1,23 @@
+#
+# Copyright 2016 Red Hat, Inc.
+#
+# Red Hat licenses this file to you under the Apache License, version
+# 2.0 (the "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.  See the License for the specific language governing
+# permissions and limitations under the License.
+#
+
+spec:
+  template:
+    spec:
+      containers:
+        - env:
+            - name: JAVA_OPTIONS
+              value: -Xmx1500m

--- a/jkube-kit/common/src/test/resources/remote-resources/sa.yml
+++ b/jkube-kit/common/src/test/resources/remote-resources/sa.yml
@@ -1,0 +1,16 @@
+#
+# Copyright (c) 2019 Red Hat, Inc.
+# This program and the accompanying materials are made
+# available under the terms of the Eclipse Public License 2.0
+# which is available at:
+#
+#     https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+#
+# Contributors:
+#   Red Hat, Inc. - initial API and implementation
+#
+
+metadata:
+  name: ribbon

